drivers/frequency/ad9528/ad9528.c

```diff
@@ -344,6 +355,417 @@ int32_t ad9528_init(struct ad9528_init_param *init_param)
 	return(0);
 }
 
+/**
+ * @brief Initialize the CLK structure.
+ *
+ * @param desc - The CLK descriptor.
+ * @param init_param - The structure holding the device initial parameters.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int ad9528_clk_init(struct no_os_clk_desc **desc,
+			   const struct no_os_clk_init_param *init_param)
+{
+	struct ad9528_dev *ad9528_d;
+
+	/* Exit if we have no init_params */
+	if (!init_param) {
+		return -EINVAL;
+	}
+
+	*desc = no_os_calloc(1, sizeof(**desc));
+	/* Exit if memory cannot be allocated */
+	if(!*desc) {
+		free(*desc);
+		return -ENOMEM;
+	}
+
+	ad9528_d = init_param->dev_desc;
+	/* Exit if no hardware device specified in init_param */
+	if(!ad9528_d) {
+		free(*desc);
+		return -ENOMEM;
+	}
+
+	(*desc)->name = init_param->name;
+	(*desc)->hw_ch_num = init_param->hw_ch_num;
+
+	(*desc)->dev_desc = (void *)no_os_calloc(1, sizeof(struct ad9528_dev));
+
+	(*desc)->dev_desc = init_param->dev_desc;
+
+	return 0;
+}
+
+/**
+ * @brief Get the current frequency of the clock.
+ *
+ * @param desc - The CLK descriptor.
+ * @param rate - The CLK rate.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad9528_clk_recalc_rate(struct no_os_clk_desc *desc,
+			       uint64_t *rate)
+{
+	struct ad9528_dev *ad9528_dev;
+	uint32_t reg_val, val;
+	uint8_t	 s_source;
+	int ret;
+
+	ad9528_dev = desc->dev_desc;
+
+	if(!ad9528_dev)
+		return -ENODEV;
+
+	s_source = ad9528_dev->pdata->channels[desc->hw_ch_num].signal_source;
+
+	val = ad9528_dev->ad9528_st.vco_out_freq[s_source];
+	if (s_source != AD9528_SYSREF) {
+		ret = ad9528_spi_read_n(ad9528_dev,
+					AD9528_CHANNEL_OUTPUT(desc->hw_ch_num),
+					&reg_val);
+		if (ret)
+			return ret;
+		val /= AD9528_CLK_DIST_DIV_REV(reg_val);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Remove the CLK structure.
+ *
+ * @param desc - The CLK descriptor.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int ad9528_clk_remove(struct no_os_clk_desc *desc)
+{
+	struct ad9528_dev *ad9528_dev;
+	int ret;
+
+	ad9528_dev = desc->dev_desc;
+
+	if(!ad9528_dev)
+		return -ENODEV;
+
+	ret = ad9528_remove(desc->dev_desc);
```

Comment:
ad9528_remove() shouldn't be here since ad9528_clk_init() didn't call the ad9528_init().

Reply:
Done.